src/VisualStudio/Core/Impl/SolutionExplorer/DiagnosticItem/BaseDiagnosticAndGeneratorItemSource.cs

```diff
@@ -7,187 +7,208 @@
 
-        public abstract AnalyzerReference? AnalyzerReference { get; }
+    [MemberNotNullWhen(true, nameof(AnalyzerReference))]
+    // Defer actual determination and computation of the items until later.
+    public bool HasItems => !_cancellationTokenSource.IsCancellationRequested;
 
-        public abstract object SourceItem { get; }
+    public IEnumerable Items => _items;
 
-        [MemberNotNullWhen(true, nameof(AnalyzerReference))]
-        public bool HasItems
-        {
-            get
-            {
-                if (_items != null)
-                {
-                    return _items.Count > 0;
-                }
-
-                if (AnalyzerReference == null)
-                {
-                    return false;
-                }
+    private async ValueTask ProcessQueueAsync(CancellationToken cancellationToken)
+    {
+        var analyzerReference = this.AnalyzerReference;
 
-                var project = Workspace.CurrentSolution.GetProject(ProjectId);
+        // If we haven't even determined which analyzer reference we're for, there's nothing to do.
+        if (analyzerReference is null)
+            return;
 
-                if (project == null)
-                {
-                    return false;
-                }
+        // If the project went away, or no longer contains this analyzer.  Shut ourselves down.
+        var project = this.Workspace.CurrentSolution.GetProject(this.ProjectId);
+        if (project is null || !project.AnalyzerReferences.Contains(analyzerReference))
+        {
+            this.Workspace.WorkspaceChanged -= OnWorkspaceChanged;
 
-                return AnalyzerReference.GetAnalyzers(project.Language).Any() ||
-                       AnalyzerReference.GetGenerators(project.Language).Any();
-            }
+            _cancellationTokenSource.Cancel();
+            _items.Clear();
+            return;
         }
 
-        public IEnumerable Items
-        {
-            get
-            {
-                if (_items == null)
-                {
-                    var project = Workspace.CurrentSolution.GetRequiredProject(ProjectId);
-                    _generalDiagnosticOption = project.CompilationOptions!.GeneralDiagnosticOption;
-                    _specificDiagnosticOptions = project.CompilationOptions!.SpecificDiagnosticOptions;
-                    _analyzerConfigOptions = project.GetAnalyzerConfigOptions();
-
-                    _items = CreateDiagnosticAndGeneratorItems(project.Id, project.Language, project.CompilationOptions, _analyzerConfigOptions);
+        var newDiagnosticItems = GenerateDiagnosticItems(project, analyzerReference);
+        var newSourceGeneratorItems = await GenerateSourceGeneratorItemsAsync(
+            project, analyzerReference).ConfigureAwait(false);
 
-                    Workspace.WorkspaceChanged += OnWorkspaceChangedLookForOptionsChanges;
-                }
+        if (_items.SequenceEqual([.. newDiagnosticItems, .. newSourceGeneratorItems]))
```

Comment:
the core logic is that once we decide we want to update, we compute what the current set of items is.  if those are the same as the last computed set, we bail out, doing nothing at all.

if the set is different in any way, we clear out the old vlaues, and add the new ones in.

Reply:
Is order significant?

Reply:
yes'ish.  We're basically saying (like before): "the children is the list of analyzer-infos followed by teh list of generator-infos".  And we compare that ordered list before/after and if it changes *at all*, then we regen.  These should very very very rarely change.  So i'm totally ok with this and nothing smart (like ignoring ordering changes).